torch/_weights_only_unpickler.py

```diff
@@ -0,0 +1,210 @@
+# Very restricted unpickler
+# Based of https://github.com/python/cpython/blob/main/Lib/pickle.py
+# Expected to be useful for loading PyTorch model weights
+# For example:
+# data = urllib.request.urlopen('https://download.pytorch.org/models/resnet50-0676ba61.pth').read()
+# buf = io.BytesIO(data)
+# weights = torch.load(buf, pickle_module=WeightsUnpickler)
+
+from collections import OrderedDict
+from pickle import (
+    APPENDS,
+    BINGET,
+    BININT,
+    BININT1,
+    BININT2,
+    BINPERSID,
+    BINPUT,
+    BINUNICODE,
+    BUILD,
+    bytes_types,
+    decode_long,
+    EMPTY_DICT,
+    EMPTY_LIST,
+    EMPTY_SET,
+    EMPTY_TUPLE,
+    GLOBAL,
+    LONG1,
+    LONG_BINGET,
+    LONG_BINPUT,
+    MARK,
+    NEWFALSE,
+    NEWOBJ,
+    NEWTRUE,
+    NONE,
+    PROTO,
+    REDUCE,
+    SETITEM,
+    SETITEMS,
+    STOP,
+    TUPLE,
+    TUPLE1,
+    TUPLE2,
+    TUPLE3,
+    UnpicklingError,
+)
+from struct import unpack
+from sys import maxsize
+from typing import Any, Dict, List
+
+import torch
+
+
+# Unpickling machinery
+
+
+class Unpickler:
+    def __init__(self, file, *, encoding: str = "UNUSED"):
+        self.readline = file.readline
+        self.read = file.read
+        self.memo: Dict[int, Any] = {}
+
+    def load(self):
+        """Read a pickled object representation from the open file.
+
+        Return the reconstituted object hierarchy specified in the file.
+        """
+        self.metastack = []
+        self.stack: List[Any] = []
+        self.append = self.stack.append
+        read = self.read
+        readline = self.readline
+        while True:
+            key = read(1)
+            if not key:
+                raise EOFError
+            assert isinstance(key, bytes_types)
+            # Risky operators
+            if key[0] == GLOBAL[0]:
+                module = readline()[:-1].decode("utf-8")
+                name = readline()[:-1].decode("utf-8")
+                full_path = f"{module}.{name}"
+                ALLOWED_GLOBALS = {
+                    "collections.OrderedDict": OrderedDict,
+                    "torch.FloatTensor": torch.FloatTensor,
+                    "torch.FloatStorage": torch.FloatStorage,
+                    "torch.LongTensor": torch.LongTensor,
+                    "torch.LongStorage": torch.FloatStorage,
+                    "torch.nn.parameter.Parameter": torch.nn.Parameter,
+                    "torch._utils._rebuild_parameter": torch._utils._rebuild_parameter,
+                    "torch._utils._rebuild_tensor_v2": torch._utils._rebuild_tensor_v2,
```

Comment:
There is also an implicit invariant, which is that each of these functions/constructors must be able to take arbitrary user input and "construct" it safely, without triggering arbitrary code execution. This is not obvious when implementing these functions and needs to be notated at all the spots.